tools/base/argsfilter/argsfilter.c

```diff
@@ -0,0 +1,183 @@
+// Copyright(c) 2018, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include "argsfilter.h"
+#include <getopt.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+
+#define RETURN_ON_ERR(res, desc)                                               \
+	do {                                                                   \
+		if ((res) != FPGA_OK) {                                        \
+			optind = 1;                                            \
+			opterr = old_opterr;                                   \
+			fprintf(stderr, "Error %s: %s\n", (desc),              \
+				fpgaErrStr(res));                              \
+			return EX_SOFTWARE;                                    \
+		}                                                              \
+	} while (0)
+
+struct config {
+	int bus;
+	int device;
+	int function;
+	int socket_id;
+} config = {.bus = -1, .device = -1, .function = -1, .socket_id = -1};
+
+#define SUPPORTED_OPTIONS 4
+
+int fpgaPropertiesFromArgs(fpga_properties filter, fpga_result *result,
+			   int *argc, char *argv[])
```

Comment:
The `fpga` prefix is usually reserved for public OPAE FPGA APIs - can we call this something else?